File: 146/pe_146.jl

using Primes

function pe_146(limit::Int64=150000000, remainder_limit::Int64=11)

  # establish some inital empty collections
  solutions = Array{Int64, 1}()
  remainders = []

  # define get_mods function
  function get_mods(p)
    mods = []
    for r in 0:p-1
      if all([(r^2 + a) % p != 0 for a in [1, 3, 7, 9, 13, 27]])
        push!(mods, r)
      end
    end
    return mods
  end

  # observe solutions mod product of initial primes must take certain values
  for r = 0:prod(primes(remainder_limit)) - 1
    if all([r % p in get_mods(p) for p in primes(remainder_limit)])
      push!(remainders, r)
    end
  end

  # define is_solution function
  function is_solution(n)
    mods = []
    for a = 1:27
      if isprime(n^2 + a)
        push!(mods, a)
      end
    end
    return mods == [1, 3, 7, 9, 13, 27]
  end

  # determine eligible candidates and check if they are solutions
  for n = 0:10:limit
    if n % (prod(primes(remainder_limit))) in remainders
      if is_solution(n)
        println(n)
        push!(solutions, n)
      end
    end
  end
  solutions = sum(solutions)
  println("Sum: $solutions")
end

@time pe_146(150000000, 13)

# note: solution for n = 150000000 is 676333270
# note: current fastest time if 253 seconds

File: 146/PE_146.py
#! venv/bin/python3.7
"""
The smallest positive integer n for which the numbers n^2+1, n^2+3, n^2+7,
n^2+9, n^2+13, and n^2+27 are consecutive primes is 10. The sum of all such
integers n below one-million is 1242490.

What is the sum of all such integers n below 150 million?

Solution: 676333270

    Note: current fastest time is 198 seconds in Julia (Python/Cython solution
    is 608 seconds)
"""
import unittest
from functools import reduce

from pyutils.miller_rabin import MillerRabin

def PE_146(limit):

    # define get_mods function
    def get_mods(p):
        mods = []
        for r in range(0, p):
            if all([(r**2 + a) % p != 0 for a in [1, 3, 7, 9, 13, 27]]):
                mods.append(r)
        return mods

    # establish inital variables
    solutions = []
    remainders = set()
    primes = [2, 3, 5, 7, 11]
    primes_product = reduce(lambda x, y : x * y, primes, 1)

    prime_checker = MillerRabin()

    # observe solutions mod product of initial primes must take certain values
    for r in range(0, primes_product):
        if all([r % p in get_mods(p) for p in primes]):
            remainders.add(r)

    # define is_solution function
    def is_solution(n):
        mods = []
        for a in range(1, 28):
            if prime_checker.is_prime(n**2 + a):
                mods.append(a)
        return mods == [1, 3, 7, 9, 13, 27]

    # determine eligible candidates and check if they are solutions
    for n in range(0, limit + 1, 10):
        if n % 1_000_000 == 0:
            print(f"  progress >>> {n / limit:2.1%}", end='\r')
        if n % primes_product in remainders:
            if is_solution(n):
                solutions.append(n)

    return sum(solutions)


class TestPE146(unittest.TestCase):
    def test_PE_146(self):
        self.assertEqual(PE_146(150000000), 676333270)


if __name__ == '__main__':
    unittest.main()
